[layouts] Atlas and reporting framework

nyalldawson opened the pull request:
Hopefully the last remaining part of layouts, apart from the composition converter work which is underway.

This resurrects atlas, and also introduces the new reporting engine. The UI for configuring reports is still a little rough, but this can be refined post freeze. The core of it should be nice and stable, and is covered by lots of unit tests.

Reports are based on the new layouts engine. They consist of multiple
nested sections. Each individual section (and the report itself)
can have an optional header and footer (which are themselves layouts,
and can consist of multiple pages!).

Two different types of sections are implemented so far:
- a standard section, which has a single, static body layout. This
can be used to embed static layouts mid way through a report
- a "field group" section, which repeats its body layout for
every feature in a layer. The features are sorted by the selected
grouping feature (with an option for ascending/descending sort).
If a field group section has child sections (e.g. another field
group section with a different field, then only features
with unique values for the group feature are iterated over.
This allows nested reports, e.g.

Report
- Country: Australia
    - State: NSW
        - Town: Sydney
        - Town: Woolongong
    - State: QLD
        - Town: Beerburrum
        - Town: Brisbane
        - Town: Emerald
- Country: NZ
    - State: ... etc

In this example country, state or town groups can have their
own headers and footers which will be inserted in the report.

Reports are configured through a new panel in the layout designer
dialog, which is shown when editing a report (created through
the Layout Manager Dialog). The organizer allows for adding
(and removing) sections to the report, and for selecting which
layout (e.g. headers, footers, bodies) to edit within the
layout designer.

nyalldawson committed:
Start working on restoring atlas

nyalldawson committed:
Add QgsLayout subclass, QgsPrintLayout

A print layout is a layout with an atlas

nyalldawson committed:
The layouts currently stored with a project are QgsPrintLayouts

In future other layout types will be stored in projects, but
for now we only have print layouts

nyalldawson committed:
Serialize atlas settings

nyalldawson committed:
Start porting atlas GUI

nyalldawson committed:
Undo/redo for atlas settings changes

nyalldawson committed:
Work on modernizing atlas

nyalldawson committed:
More work on atlas GUI

nyalldawson committed:
Atlas sort can be an expression

nyalldawson committed:
Remove outdated TODO

nyalldawson committed:
Use messagebar for atlas messages instead of message box

nyalldawson committed:
Add some unit tests

nyalldawson committed:
Restore atlas based autogenerated filenames

nyalldawson committed:
Add atlas expression context scope to layouts

nyalldawson committed:
Working label updates on atlas feature change

nyalldawson committed:
Restore attribute table atlas handling

nyalldawson committed:
Restore shape atlas handling

nyalldawson committed:
Restore some more atlas handling

nyalldawson committed:
Restore some more atlas style related UI

nyalldawson committed:
Restore HTML item atlas functionality

nyalldawson committed:
Restore picture atlas handling

nyalldawson committed:
Code shuffle, to make QgsLayoutContext aware of parent QgsLayout

nyalldawson committed:
Port current geometry from atlas to layouts

nyalldawson committed:
Restore legend atlas behavior

nyalldawson committed:
Remove some outdated todos

nyalldawson committed:
Fix some untranslatable strings

nyalldawson committed:
Remove outdated icon support for item metadata

Now handled in QgsLayoutItem directly

nyalldawson committed:
Fix some map item todos

nyalldawson committed:
Port predefined scale handling to layouts

nyalldawson committed:
Restore atlas map handling

nyalldawson committed:
Tighten QgsAbstractLayoutIterator interface

nyalldawson committed:
Flesh out QgsAbstractLayoutIterator interface

nyalldawson committed:
Add api to export layout atlas as images

nyalldawson committed:
Hookup raster atlas export in gui

nyalldawson committed:
Restore atlas raster format handling

nyalldawson committed:
Restore svg atlas export

nyalldawson committed:
Restore atlas single file pdf export

nyalldawson committed:
Restore atlas multiple pdf exports

nyalldawson committed:
Force a refresh after disabling view updates

nyalldawson committed:
Fix broken atlas handling of z/m layer types

nyalldawson committed:
Fix some layout expression context handling, restore some tests

nyalldawson committed:
Restore some more atlas related tests

nyalldawson committed:
Restore more atlas tests

nyalldawson committed:
Restore more unit tests

nyalldawson committed:
Try (and fail) to avoid a qApp->processEvents() call

I just can't find any other reliable way to wait until javascript
execution in a web page has completed.

nyalldawson committed:
Expand exporter tests

nyalldawson committed:
Ensure that every iteration through a layout iterator only
references the current iteration layout

nyalldawson committed:
Refactor layout context

Split render context from reporting context

nyalldawson committed:
Add a clone method to QgsLayout

nyalldawson committed:
Remove const from count method

nyalldawson committed:
[FEATURE] Reporting framework

Reports are based on the new layouts engine. They consist of multiple
nested sections. Each individual section (and the report itself)
can have an optional header and footer (which are themselves layouts,
and can consist of multiple pages!).

Two different types of sections are implemented so far:
- a standard section, which has a single, static body layout. This
can be used to embed static layouts mid way through a report
- a "field group" section, which repeats its body layout for
every feature in a layer. The features are sorted by the selected
grouping feature (with an option for ascending/descending sort).
If a field group section has child sections (e.g. another field
group section with a different field, then only features
with unique values for the group feature are iterated over.
This allows nested reports, e.g.

Report
- Country: Australia
    - State: NSW
        - Town: Sydney
        - Town: Woolongong
    - State: QLD
        - Town: Beerburrum
        - Town: Brisbane
        - Town: Emerald
- Country: NZ
    - State: ... etc

In this example country, state or town groups can have their
own headers and footers which will be inserted in the report.

Reports are configured through a new panel in the layout designer
dialog, which is shown when editing a report (created through
the Layout Manager Dialog). The organizer allows for adding
(and removing) sections to the report, and for selecting which
layout (e.g. headers, footers, bodies) to edit within the
layout designer.

nyalldawson committed:
Add missing factory annotation

nyalldawson committed:
Add report section subclass with single layout as body

nyalldawson committed:
Add reprt iteration test

nyalldawson committed:
Hookup filename generation for reports

nyalldawson committed:
Add a unit test for exporting reports

nyalldawson committed:
Correct project and parent handling for report sections

nyalldawson committed:
Implement QgsReportSectionFieldGroup

nyalldawson committed:
Code shuffle and cleanup

nyalldawson committed:
Add a common base class for print layouts and reports, and adapt
QgsLayoutManager to suit

nyalldawson committed:
Implement serialization of reports

nyalldawson committed:
Implement sort order customisation for field groups

nyalldawson committed:
Protect layout view against null layouts

nyalldawson committed:
Fix multiple empty layouts created when reloading project

In related news... I find it frustrating that there's no
way in Qt to restrict elementsByTagName to direct descendants
only, and that all other available API calls for searching
only direct descendants are much more fiddly...

nyalldawson committed:
Add button in layout manager dialog to create a new report

nyalldawson committed:
Disable layout designer actions when no layout is set

nyalldawson committed:
Cleaner way to hide atlas controls when not using a print layout

nyalldawson committed:
Add crappy inappropriate icons for layout types

nyalldawson committed:
Sort layouts in layout manager

nyalldawson committed:
Fix clashing inherited methods

nyalldawson committed:
Add a row method to QgsAbstractReportSection

nyalldawson committed:
Fix project retrieval for QgsReport

nyalldawson committed:
Start on ui for configuring reports

nyalldawson committed:
Add buttons for exporting reports

nyalldawson committed:
More UI work on reports

nyalldawson committed:
Fix progress bar with report export

nyalldawson committed:
Expand report sections by default

nyalldawson committed:
Fix missing Report top level item in organizer widget

nyalldawson committed:
Auto select new report sections after adding them

nyalldawson committed:
Dox and spelling

nyalldawson committed:
Add test mask

nirvn commented:
I've tested the atlas part of this PR, switching between features and exporting {images, pdf, svg} atlases worked like a charm. The two tiny issues I spotted were:
- main canvas not refreshed when an atlas feature is changed in an open layout (it requires the user to do a manual refresh or pan/zoom around to see the changed symbology)
- select atlas feature action (through main window's canvas) is missing

Great to see this being so close to the finish line.

nyalldawson committed:
Ensure main canvas is refreshed when atlas preview feature changes

nyalldawson committed:
Fix crash on close project with layout designer open

nyalldawson committed:
Add new icon for atlas exports (courtesy of @nirvn)

nyalldawson committed:
Add method to seek atlas directly to a QgsFeature

nyalldawson committed:
Swap 'set atlas feature' action from compositions to layouts

Time to start making a permenant switch...

nyalldawson commented:
@nirvn thanks for the testing - both issues are now fixed

timlinux commented:
Ah thanks for this - ignore my question on loomio if atlas is coming in 3.0 then :-P

nyalldawson committed:
Restore direct print actions

nirvn commented:
Just tested the new report feature; some UI/UX comments:
- assuming that map items can be framed using a given report feature, the [x] controlled by atlas should be renamed [x] controlled by report in this context
- IMHO, double-clicking on a report section should open its layout (i.e, same as when we’re clicking on the section body’s [ edit ] button)
- when creating a new report, the dialog’s string should be renamed to not refer to a unique “print layout” title
- when you hide the layout sheet(s) when exporting a report (as well as an atlas), you should also make sure you deselect item(s) – I had a map item selected, with its item properties page open, and I could see the extent value changes as the report / atlas was being exported (I assume I could also interact with it and mess things up :) )
- by default, the layout manager dialog’s bottom row of buttons have labels chopped off:
- is there any way to move report sections around? (I assumed drag’n dropping would do, but it doesn’t):
![screenshot from 2018-01-02 09-50-00](https://user-images.githubusercontent.com/1728657/34473670-31b0798c-efa8-11e7-89eb-ed6e53978a50.png)


More generic comments on the report (disclosure: I’m not sure I fully understand how report’s field groups work :) ):
- On the sub-field group, it doesn’t seem to work (or at least not as I am expecting it to). If I have a layer with districts, and I want to re-group them by parent province, I am using a province_code field for the district layers as a parent field group, and a district_code field as a sub field group. When I export the report, I expect the parent field group to cycle through unique province_code, and the sub-field group to cycle through the unique district_code that matches a given province_code. Yet, when I export, the parent province_code repeats itself (i.e. it simply cycles through all districts), and then exports all district_code features within a given province_code, which ends up repeating a given province_code set of districts again and again.
- Let’s say I have a polygon layer with nationwide districts, and I want to use one field (say province_id) to show the overall nb of districts, as well as a sub field group (say district_id) to show individual districts within a given province_id value. Right now, when I have a section body set for the province_id field group, there’s no way I can set a map item to show all features with a given province_id value.

nyalldawson committed:
Fix missing feature contexts for report section header/footers

and expand unit tests

nyalldawson commented:
> is there any way to move report sections around? (I assumed drag’n dropping would do, but it doesn’t):

nope ... 3.2 feature ;)

> On the sub-field group, it doesn’t seem to work (or at least not as I am expecting it to). If I have a layer with districts, and I want to re-group them by parent province, I am using a province_code field for the district layers as a parent field group, and a district_code field as a sub field group. When I export the report, I expect the parent field group to cycle through unique province_code, and the sub-field group to cycle through the unique district_code that matches a given province_code. Yet, when I export, the parent province_code repeats itself (i.e. it simply cycles through all districts), and then exports all district_code features within a given province_code, which ends up repeating a given province_code set of districts again and again.

Should be fixed now

> Let’s say I have a polygon layer with nationwide districts, and I want to use one field (say province_id) to show the overall nb of districts, as well as a sub field group (say district_id) to show individual districts within a given province_id value. Right now, when I have a section body set for the province_id field group, there’s no way I can set a map item to show all features with a given province_id value.

Aggregate geometries aren't created for field groups - that's a potential 3.2 feature. In the meantime you can use an expression function to handle this.

nyalldawson commented:
> when creating a new report, the dialog’s string should be renamed to not refer to a unique “print layout” title

can you clarify?

nyalldawson committed:
Use 'report' text instead of 'atlas' when inside a report designer

nyalldawson commented:
> assuming that map items can be framed using a given report feature, the [x] controlled by atlas should be renamed [x] controlled by report in this context

Fixed

nyalldawson commented:
> IMHO, double-clicking on a report section should open its layout (i.e, same as when we’re clicking on the section body’s [ edit ] button)

I disagree - there's many cases when you don't want a body (e.g. field group sections with children), and this behavior would likely result in many accidental body creations.

nyalldawson committed:
Nicer ui widgets for sections, add missing control for
disabling body sections

nyalldawson commented:
@nirvn - thanks for the testing! I've just updated based on our convo

nirvn commented:
@nyalldawson , compiling the updated branch.

Two tiny UX bummers:
- when creating a new report, the report panel should be automatically focused, and opened if previously closed. On my machine, a new report opens with a report panel closed or unfocused, and that's not helpful :)
- when opening an header / footer / section body and dragging a new item onto the sheet, the item properties panel should automatically focus. Right now, when creating a report section, I end up having to manually focus the item properties and that's again just a bit of an irritant we can easily deal with auto focusing behavior.

Also, IMHO, we need a way to either disable and/or remove a section body; my preference would be a [x] section body checkbox just like header and footer to enable / disable the section body.

nirvn commented:
@nyalldawson , also, when no item is focused in the report section list (it happens for e.g. after you delete an existing item), adding a new section should append it to the root node instead of failing altogether. I've hit this a couple of times while testing earlier.

nirvn commented:
> Also, IMHO, we need a way to either disable and/or remove a section body; my preference would be a [x] section body checkbox just like header and footer to enable / disable the section body.

Cool, I see you've added it already. Brilliant. Much better UI logic.

nirvn commented:
> when creating a new report, the dialog’s string should be renamed to not refer to a unique “print layout” title

> can you clarify?

I'd remove the word "print" since that dialog is used to both create actual print layouts, as well as report layouts.

nyalldawson committed:
Fix failing unit test

timlinux commented:
I think I am missing something basic here or there is a mac issue at play. When I try to create a report I don't get any kind of preview on the layout and the edit buttons don't seem to do anything:

![licecap](https://user-images.githubusercontent.com/178003/34487823-a5d94970-efde-11e7-864e-2bffb1ea8c6c.gif)

timlinux commented:
I also got a crash when trying to remove the top level report section

nyalldawson commented:
> I also got a crash when trying to remove the top level report section

That's not possible to... There's a guard against it (which obviously isn't working correctly!)

nyalldawson commented:
> I think I am missing something basic here or there is a mac issue at play. When I try to create a report I don't get any kind of preview on the layout and the edit buttons don't seem to do anything:

We don't have an interactive preview for reports.like we do for atlas (hopefully we can do this in the next round of funding). But I suspect it's workijg correctly, just different to your expectations.

All report sections, including the heades, footers, and bodies are themselves layouts. They consist of 1-n pages of the standard layout items (maps/labels etc)

What you see when you hit "edit" is just THAT layout in the designer. You won't see the full report until you export it. 

Try adding some labels etc to the various sections and I think it'll become clearer.

timlinux commented:
@nyalldawson thanks for the explanation - makes much more sense now. I got a wierd thing on Mac exporting the report - the progress seems to stay at 100% all the time:

![licecap](https://user-images.githubusercontent.com/178003/34522191-2bb796ee-f09a-11e7-8867-42f4084acf99.gif)

timlinux commented:
I am also getting an issue on MacOS if I delete text inside of the label item properties, it deletes the whole label:

![licecap](https://user-images.githubusercontent.com/178003/34529224-f34ec16e-f0b3-11e7-8f12-88a52fa6af94.gif)

timlinux commented:
Another issue I found - after selecting a font in MacOS, the focus returns to the main QGIS window instead of the composer window:

![licecap](https://user-images.githubusercontent.com/178003/34533530-47423be8-f0c3-11e7-8364-ad9642c099d5.gif)

nyalldawson commented:
> Another issue I found - after selecting a font in MacOS, the focus returns to the main QGIS window instead of the composer window:

That'll apply everywhere the new font button is used. Can you open an issue about this? (It's unrelated to this pr)

nyalldawson commented:
> I am also getting an issue on MacOS if I delete text inside of the label item properties, it deletes the whole label:

Open an issue too :p (it's not this pr). Also test with composer please and see if it's related to the new designer or is a qt change/issue.

nyalldawson commented:
> I got a wierd thing on Mac exporting the report - the progress seems to stay at 100% all the time:

It should be using the "unknown" total progress type bar - i.e. on Linux/win the one which bounces back and forth but doesn't fill. Maybe OSX doesn't render this unknown progress bar? (You could test with a QProgressBar with 0 min and max)

timlinux commented:
> That'll apply everywhere the new font button is used. Can you open an issue about this? (It's unrelated to this pr)

OK filed here: https://issues.qgis.org/issues/17783

> Label delete issue: Open an issue too :p (it's not this pr). Also test with composer please and see if it's related to the new designer or is a qt change/issue.

OK filed here: https://issues.qgis.org/issues/17784

> It should be using the "unknown" total progress type bar - i.e. on Linux/win the one which bounces back and forth but doesn't fill. Maybe OSX doesn't render this unknown progress bar? (You could test with a QProgressBar with 0 min and max)

I guess I was just expecting that we would have an idea of how many pages are being generated a-priori  and thus be able to show the progress accordingly.

elpaso commented:
Hi, I enabled the atlas but the "Controlled by Atlas"  map extent options remains uncheckable
![qgis-atlas-controlled-by](https://user-images.githubusercontent.com/142164/34558255-8dc7a20e-f13d-11e7-8234-16431d681c80.png)

elpaso commented:
The "Preview atlas" item in the "Atlas" menu should IMHO be checkable because it changes the view mode.

![qgis-atlas-view-mode](https://user-images.githubusercontent.com/142164/34558296-cc0bd486-f13d-11e7-9c98-64e1d43ea52b.png)

elpaso commented:
I'm not sure if this is the expected behavior but when you first create a report from the layout manager, the windows opens without any page elements: this is confusing, I had to click on the header to see the page:

![qgis-report](https://user-images.githubusercontent.com/142164/34559857-ddb4d790-f143-11e7-8ab7-55adda41817a.gif)

elpaso commented:
Still on reports, when you click on "Edit" button after "Include body" whatever is in the current page is deleted: is this expected?

elpaso commented on src/app/layout/qgslayoutmanagerdialog.cpp:
```diff
@@ -285,6 +292,31 @@ void QgsLayoutManagerDialog::mTemplatesUserDirBtn_pressed()
+  report->setHeaderEnabled( true );
+
+  QgisApp::instance()->openLayoutDesignerDialog( report.get() );
+  QgsProject::instance()->layoutManager()->addLayout( report.release() );
```
I don't see what's the advantage in using a unique_ptr here that we are releasing it and there are no multiple return points neither exceptions, but if this is the modern C++ way, let's follow it!

nyalldawson replied:
That's what I'm currently leaning toward - even if there's currently no other return points, it doesn't mean the code will always be this way. So in future if someone refactors, they'll have no choice but to respect the ownership of the pointer. It also makes it very clear when reading the code who has ownership - with Qt's parent system it can be difficult when reading over code to know exactly who has ownership of an object, and whether there's a leak or something is correctly parented. Plus, smarter minds then me have declared that modern c++ practice is to *never* use new/delete, and I'm happy to go along with them :) *


* we can't do this always, because of Qt parenting system and also because sip has no concept of smart pointers yet.

elpaso commented on src/app/layout/qgslayoutmanagerdialog.cpp:
```diff
@@ -63,7 +65,9 @@ QgsLayoutManagerDialog::QgsLayoutManagerDialog( QWidget *parent, Qt::WindowFlags
 
```
I realize that it's not in this PR but I'd like to see more `QgsSettings::Sections` instead of hardcoded prefixes like `UI`

nyalldawson replied:
Fixed

elpaso commented on tests/src/core/testqgslayoutlabel.cpp:
```diff
@@ -44,6 +47,7 @@ class TestQgsLayoutLabel : public QObject
     void evaluation();
     // test expression evaluation when a feature is set
     void feature_evaluation();
+    void feature_evaluation2();
```
Maybe rename to make it clearer that one test is using atlas and the other one is not

nyalldawson replied:
Fixed

elpaso commented:
I tested briefly the Html frame, I probably missed something but when "Extend to the next page" is on I couldn't find a way to set margins or to keep the original frame size on all pages after the first. I'm not sure what would be the desired outcome in this case, but an option to set the same frame size/position on the pages > 1 would be probably useful.

![qgis-htmlframe](https://user-images.githubusercontent.com/142164/34574785-c89974c6-f178-11e7-9104-df65415ff300.png)

nyalldawson committed:
Use correct layout type (report/print layout) in new title dialog

nyalldawson commented:
> when creating a new report, the dialog’s string should be renamed to not refer to a unique “print layout” title

Fixed

nyalldawson commented:
> when you hide the layout sheet(s) when exporting a report (as well as an atlas), you should also make sure you deselect item(s) – I had a map item selected, with its item properties page open, and I could see the extent value changes as the report / atlas was being exported (I assume I could also interact with it and mess things up :) )

I'm not sure this is an issue - the progress dialogs are modal and block the layout designer, so it's not possible to change anything during the export. Might address this in a future PR if it proves to be an issue.

nyalldawson commented:
> by default, the layout manager dialog’s bottom row of buttons have labels chopped off:

I'd love a total redesign of this dialog, as I don't consider it very nice UX in general. But I'm also out of nice ideas for a rearrange. Got any ideas for me?

nyalldawson committed:
Ensure report dock is visible when opening a report designer

nyalldawson commented:
> when creating a new report, the report panel should be automatically focused, and opened if previously closed. On my machine, a new report opens with a report panel closed or unfocused, and that's not helpful :)

Fixed

nyalldawson committed:
Make report/atlas settings actions checkable, so they can also hide the panels

nyalldawson commented:
> when opening an header / footer / section body and dragging a new item onto the sheet, the item properties panel should automatically focus. Right now, when creating a report section, I end up having to manually focus the item properties and that's again just a bit of an irritant we can easily deal with auto focusing behavior.

@nirvn 
That's a "wont-fix" from me. I personally like to work with compositions in a full page view, with the docks hidden (via the ctrl tab shortcut). I'd find it annoying if the item properties dock kept reopening and refocusing when clicking items in the layout then.

I've tried to mitigate the original issue somewhat by moving the report settings dock to the left by default, so that it's not grouped with the item properties dock tab.

nyalldawson committed:
Don't tabify report settings with item properties

nyalldawson committed:
If no section is selected, add new sections to report itself

nyalldawson committed:
Fix crash on report designer close

nyalldawson commented:
>  also, when no item is focused in the report section list (it happens for e.g. after you delete an existing item), adding a new section should append it to the root node instead of failing altogether. I've hit this a couple of times while testing earlier.

 Fixed

nyalldawson commented:
> I also got a crash when trying to remove the top level report section

@timlinux  I can't reproduce this - but I've changed the widget to explicitly disable the remove button when no child section is selected.

nyalldawson committed:
Explicitly disable remove button if no child report section is selected

nyalldawson commented:
@elpaso 
> I mean that a real checkbox would be probably more noticeable.

Isn't this a theme issue? I see the same on Fedora, but for all checkable menu actions (including the in main canvas window)

nyalldawson committed:
Auto generated names for reports are 'Report #'

nyalldawson committed:
Make some actions apply to reports when a report designer is open

nyalldawson commented:
@elpaso 

> I tested briefly the Html frame, I probably missed something but when "Extend to the next page" is on I couldn't find a way to set margins or to keep the original frame size on all pages after the first. I'm not sure what would be the desired outcome in this case, but an option to set the same frame size/position on the pages > 1 would be probably useful.

Honestly... I can't see the use case for this mode either and would love to remove it. But it's been in QGIS since before 2.0 I think, and I'm reluctant to remove it without fully understanding what the original driver for this feature was.

In my opinion the correct way to handle these "sometimes multipage" items is to manually create the additional pages, with whatever titles and labels and other elements you desire on the following pages, and then use manual frames with the "don't export page if frame is empty" option checked.

nyalldawson committed:
Fix atlas actions not immediately available for map items
when toggling atlas enabled

nyalldawson commented:
@elpaso 

>  I enabled the atlas but the "Controlled by Atlas" map extent options remains uncheckable

Fixed

nyalldawson committed:
Rename some test methods

nyalldawson committed:
Use correct QgsSettings keys for layouts

nyalldawson committed:
Show section name in layout designer title

To make it clearer which section is currently being edited

nyalldawson committed:
Add icons for report section types, and show pencil 'editing'
icon in report organizer for section currently being edited in
the designer

This should help indicate to users which section is currently being
edited and give more visual hints as to exactly what's happening
in the ui.

nirvn commented:
Fantastic set of follow up commits. The icon + pencil commit greatly improves the UX.

nyalldawson commented:
> Fantastic set of follow up commits. The icon + pencil commit greatly improves the UX.

I'm thinking of also adding a label which floats above the first page too, just to make it super-clear which section is shown in the designer. But I'll merge this first.

nirvn commented:
Oh if you have the energy to implement the floating label, that'd be great too. And yeah, this PR is ready to be merge.

Can't wait for my CPU to stop "wasting" cycles compiling the composer stuff once the migration PR is merged.
